spec/models/form_item_spec.rb

```diff
@@ -0,0 +1,22 @@
+require "spec_helper"
+
+describe FormItem do
+  before do
+    @user = create(:user, role_name: 'coordinator')
+    @form = create(:form, question_types: ['text', ['text', 'text']])
+    @qing = @form.c[0]
+    @qing_group = @form.c[1]
+  end
+
+  describe "check_ancestry_integrity" do
+    it "should return true" do
+      expect(@qing.check_ancestry_integrity(@qing_group.id)).to be_truthy
+    end
+
+    it "should return false" do
+      @qing_group.ancestry = @qing.id
+      @qing_group.save
+      expect(@qing.check_ancestry_integrity(@qing_group.id)).to be_falsy
+    end
+  end
+end
```

Comment:
I would like to see some tests here that really challenge the system:
1. If you move an item down, do the ranks adjust properly? Same if you move an item up.
1. If you change an item's parent, what happens to its rank?
1. If you change an item's rank and that invalidates a condition, what happens? (See Condition.verify_ordering method).
1. Each of these things should be tried both on the top level and on a sublevel.

I think if you write these tests you might find that there are bugs since we are using those old methods.